client/packages/programs/src/JsonForms/components/ProgramEvent.tsx

```diff
@@ -0,0 +1,227 @@
+        }),
+      ])
+      .optional(),
+    /** Show a program events that target documents of a certain type.
+     * Note, this only includes events that are not targeting a specific document name.
+     * If not specified the program events for the current document are displayed
+     * (using the document name of the current JSONForms document)
+     */
+    documentType: z.string().optional(),
+    eventType: z.string(),
+    /**
+     * Display option based on type.
+     */
+    display: z
+      .discriminatedUnion('type', [
+        z.object({ type: z.literal('number') }),
+        z.object({
+          type: z.literal('string'),
+          multiline: z.boolean().optional(),
+          rows: z.number().optional(),
+          show: z.array(
+            z
+              .tuple([z.string(), z.string().optional()])
+              .rest(z.string().optional())
+          ),
+        }),
+      ])
+      .optional(),
+  })
+  .strict();
+type Options = z.infer<typeof Options>;
+
+const extractAt = (
+  encounter?: EncounterFragment,
+  program?: ProgramEnrolmentFragment,
+  options?: Options
+): Date => {
+  const date = new Date();
+
+  if (options?.at?.type === 'encounter') {
+    switch (options.at.encounterStartDatetime) {
+      case 'before':
+        const before = new Date(
+          new Date(encounter?.startDatetime ?? date).getTime() - 1
+        );
+        return before;
+      case 'after':
+        return new Date(encounter?.startDatetime ?? date);
+      default:
+        // eslint-disable-next-line @typescript-eslint/no-unused-vars
+        ((_: never) => {})(options.at.encounterStartDatetime as never);
+    }
+  } else if (options?.at?.type === 'programEnrolment') {
+    switch (options.at.programEnrolmentDatetime) {
+      case 'before':
+        const before = new Date(
+          new Date(program?.enrolmentDatetime ?? date).getTime() - 1
+        );
+        return before;
+      case 'after':
+        return new Date(program?.enrolmentDatetime ?? date);
+      default:
+        // eslint-disable-next-line @typescript-eslint/no-unused-vars
+        ((_: never) => {})(options.at.programEnrolmentDatetime as never);
+    }
+  }
+
+  return date;
+};
+
+const getDisplayOptions = (
+  data: string | null | undefined,
+  options?: Options
+) => {
+  const show =
+    options?.display?.type === 'string' ? options?.display?.show : null;
+
+  if (!show) {
+    return data;
+  }
+
+  const displayValue = show.find(value => value[0] === data)?.[1];
+  return displayValue ?? '';
+};
+
+const UIComponent = (props: ControlProps) => {
+  const { label, uischema, config } = props;
+  const patientId = config?.patientId;
+  const [datetime, setDatetime] = useState<Date | undefined>();
+
+  const { data: encounter } = useEncounter.document.byDocName(
+    config.documentName
+  );
+  const { data: program } = useProgramEnrolments.document.byDocName(
+    config.documentName
+  );
```

Comment:
Nit: you could probably avoid an extra api call here by enabling/disabling the react-query hook depending on the options but I think its good enough for now...